aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -228,7 +228,9 @@ public void deleteFiles(Iterable<String> paths) throws BulkDeletionFailureExcept
           throw new RuntimeException("Interrupted when waiting for deletions to complete", e);
         }
       }
-
+      if (executorService() != null) {
```

Comment:
Actually, Shutting down the executor service here is not appropriate because a client might create a single S3fileIO object and call multiple methods on it. For instance one can call the deleteFiles method multiple times on the same S3FileIo object. So, if we shut down the executor service then the second deleteFiles method will fail. 

You can actually add the following line to the end of this snippet `executorService = null` . That way, if anyone calls the deleteFiles method method multiple times on the same S3FileIo object a new ExecutorService will be recreated. But that would nullify the need for executorService method. I think the point of having that method is to reuse the executorService.

I think it might actually be better to remove this snippet entirely in this case and let the garbage collector take care of shutting down the executor service. But that would probably not solve the issue that you are facing. 

You can maybe add a method like resetExecutorService and call it in your fork. Or you can maybe make executorService a static method, so that the same thread pool is reused every time and you wouldn't see too many threads created at the same time. But that comes with its own problem. The awsProperties object which is needed to create the executorService object is dependent on the S3FileIo object not the class. So, that would be a problem.

Reply:
Then I think it is better to add a method in FileIO that can pass in a custom executorService to execute the Delete method. We can also close the threads pool of Procedure uniformly in the upper layer